core/services/gateway/handlers/webapicapabilities/handler.go

```diff
@@ -1,6 +1,156 @@
+
 const (
 	// NOTE: more methods will go here. HTTP trigger/action/target; etc.
 	MethodWebAPITarget = "web_api_target"
 )
+
+type handler struct {
+	don             handlers.DON
+	lggr            logger.Logger
+	httpClient      network.HTTPClient
+	nodeRateLimiter *common.RateLimiter
+}
+
+type HandlerConfig struct {
+	NodeRateLimiter common.RateLimiterConfig `json:"nodeRateLimiter"`
+}
+
+var _ handlers.Handler = (*handler)(nil)
+
+func NewHandler(handlerConfig json.RawMessage, donConfig *config.DONConfig, don handlers.DON, httpClient network.HTTPClient, lggr logger.Logger) (*handler, error) {
+	var cfg HandlerConfig
+	err := json.Unmarshal(handlerConfig, &cfg)
+	if err != nil {
+		return nil, err
+	}
+	nodeRateLimiter, err := common.NewRateLimiter(cfg.NodeRateLimiter)
+	if err != nil {
+		return nil, err
+	}
+
+	return &handler{
+		don:             don,
+		lggr:            lggr.Named("WebAPIHandler." + donConfig.DonId),
+		httpClient:      httpClient,
+		nodeRateLimiter: nodeRateLimiter,
+	}, nil
+}
+
+func (h *handler) HandleUserMessage(ctx context.Context, msg *api.Message, callbackCh chan<- handlers.UserCallbackPayload) error {
+	return nil
+}
+
+// sendHTTPMessageToClient is an outgoing message from the gateway to external endpoints
+// returns message to be sent back to the capability node
+func (h *handler) sendHTTPMessageToClient(ctx context.Context, req network.HTTPRequest, msg *api.Message) (*api.Message, error) {
+	var payload TargetResponsePayload
+	resp, err := h.httpClient.Send(ctx, req)
+	if err != nil {
+		return nil, err
+	} else {
+		payload = TargetResponsePayload{
+			ExecutionError: false,
+			StatusCode:     uint16(resp.StatusCode),
+			Headers:        resp.Headers,
+			Body:           resp.Body,
+		}
+	}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+	return &api.Message{
+		Body: api.MessageBody{
+			MessageId: msg.Body.MessageId,
+			Method:    msg.Body.Method,
+			DonId:     msg.Body.DonId,
+			Payload:   payloadBytes,
+		},
+	}, nil
+}
+
+func (h *handler) handleWebAPITargetMessage(ctx context.Context, msg *api.Message, nodeAddr string) error {
+	h.lggr.Debugw("handling web api target message", "messageId", msg.Body.MessageId, "nodeAddr", nodeAddr)
+	if !h.nodeRateLimiter.Allow(nodeAddr) {
+		return fmt.Errorf("rate limit exceeded for node %s", nodeAddr)
+	}
+	var targetPayload TargetRequestPayload
+	err := json.Unmarshal(msg.Body.Payload, &targetPayload)
+	if err != nil {
+		return err
+	}
+	// send message to target
+	req := network.HTTPRequest{
+		Method:  targetPayload.Method,
+		URL:     targetPayload.URL,
+		Headers: targetPayload.Headers,
+		Body:    targetPayload.Body,
+		Timeout: time.Duration(targetPayload.TimeoutMs) * time.Millisecond,
+	}
+	// this handle method must be non-blocking
+	// send response to node (target capability) async
+	// if there is a non-HTTP error (e.g. malformed request), send payload with success set to false and error messages
+	go func() {
```

Comment:
Is this how other handlers ensure non-blocking requests?

There's a few things that feel smelly here: 1) the untracked goroutine, and 2) the disconnected context

If anything upstream were to cancel the context, that could feed into cancelling the request too, which doesn't feel right.

Reply:
there is one other handler in functions but it doesn't make any external networking calls so it's not implemented as a async task like this. Just for context, the non-blocing pattern using goroutine was suggested here - https://github.com/smartcontractkit/chainlink/pull/14448#discussion_r1762125129

Although the goroutines are untracked, the goroutine here should always terminate eventually because context with timeout is passed to both `SendToNode` and `sendHTTPMessageToClient`. I don't think I understood the concern around disconnected context here. If upstream (gateway service) cancels the context, then the request from target can eventually be cancelled and target capability will timeout resulting in workflow failure, which seems ok.